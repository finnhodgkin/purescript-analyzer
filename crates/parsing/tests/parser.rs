macro_rules! parser_tests {
    ($($name:ident => $source:expr),+ $(,)?) => {
        $(
            #[test]
            fn $name() {
                let lexed = lexing::lex($source);
                let tokens = lexing::layout(&lexed);
                let (node, errors) = parsing::parse(&lexed, &tokens);
                insta::assert_debug_snapshot!((node, errors));
            }
        )+
    };
}

parser_tests!(
    comment_pre_post => include_str!("parser/CommentPrePost.purs"),
    lexer_error_inserted => include_str!("parser/LexerErrorInserted.purs"),
    module_export => include_str!("parser/ModuleExport.purs"),
    module_export_class => include_str!("parser/ModuleExportClass.purs"),
    module_export_class_error_end => include_str!("parser/ModuleExportClassErrorEnd.purs"),
    module_export_comma_error => include_str!("parser/ModuleExportCommaError.purs"),
    module_export_data => include_str!("parser/ModuleExportData.purs"),
    module_export_data_comma_error => include_str!("parser/ModuleExportDataCommaError.purs"),
    module_export_data_item_after_double_period => include_str!("parser/ModuleExportDataItemAfterDoublePeriod.purs"),
    module_export_data_item_before_double_period => include_str!("parser/ModuleExportDataItemBeforeDoublePeriod.purs"),
    module_export_data_item_invalid => include_str!("parser/ModuleExportDataItemInvalid.purs"),
    module_export_empty => include_str!("parser/ModuleExportEmpty.purs"),
    module_export_invalid => include_str!("parser/ModuleExportInvalid.purs"),
    module_export_type_operator => include_str!("parser/ModuleExportTypeOperator.purs"),
    module_export_type_operator_error => include_str!("parser/ModuleExportTypeOperatorError.purs"),
    module_export_type_operator_error_end => include_str!("parser/ModuleExportTypeOperatorErrorEnd.purs"),
    module_header => include_str!("parser/ModuleHeader.purs"),
    module_header_prefixed => include_str!("parser/ModuleHeaderPrefixed.purs"),
);
