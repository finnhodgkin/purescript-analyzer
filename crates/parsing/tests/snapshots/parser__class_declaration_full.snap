---
source: crates/parsing/tests/parser.rs
expression: "(node, errors)"
snapshot_kind: text
---
(
    Module@0..208
      ModuleHeader@0..35
        MODULE@0..6 "module"
        Comment@6..7
          WHITESPACE@6..7 " "
        ModuleName@7..28
          UPPER@7..27 "ClassDeclarationFull"
          Comment@27..28
            WHITESPACE@27..28 " "
        ModuleExportList@28..28
        WHERE@28..33 "where"
        Comment@33..35
          WHITESPACE@33..35 "\n\n"
      LAYOUT_START@35..35 ""
      ModuleImports@35..35
      ModuleStatements@35..208
        ClassDeclaration@35..116
          CLASS@35..40 "class"
          Comment@40..41
            WHITESPACE@40..41 " "
          ClassConstraints@41..59
            TypeApplicationChain@41..56
              TypeConstructor@41..52
                QualifiedName@41..52
                  UPPER@41..51 "TypeEquals"
                  Comment@51..52
                    WHITESPACE@51..52 " "
              TypeVariable@52..54
                LOWER@52..53 "b"
                Comment@53..54
                  WHITESPACE@53..54 " "
              TypeVariable@54..56
                LOWER@54..55 "a"
                Comment@55..56
                  WHITESPACE@55..56 " "
            LEFT_THICK_ARROW@56..58 "<="
            Comment@58..59
              WHITESPACE@58..59 " "
          ClassHead@59..74
            UPPER@59..69 "TypeEquals"
            Comment@69..70
              WHITESPACE@69..70 " "
            TypeVariable@70..72
              LOWER@70..71 "a"
              Comment@71..72
                WHITESPACE@71..72 " "
            TypeVariable@72..74
              LOWER@72..73 "b"
              Comment@73..74
                WHITESPACE@73..74 " "
          ClassFunctionalDependencies@74..91
            PIPE@74..75 "|"
            Comment@75..76
              WHITESPACE@75..76 " "
            FunctionalDependencyDetermines@76..82
              LOWER@76..77 "a"
              Comment@77..78
                WHITESPACE@77..78 " "
              RIGHT_ARROW@78..80 "->"
              Comment@80..81
                WHITESPACE@80..81 " "
              LOWER@81..82 "b"
            COMMA@82..83 ","
            Comment@83..84
              WHITESPACE@83..84 " "
            FunctionalDependencyDetermines@84..91
              LOWER@84..85 "b"
              Comment@85..86
                WHITESPACE@85..86 " "
              RIGHT_ARROW@86..88 "->"
              Comment@88..89
                WHITESPACE@88..89 " "
              LOWER@89..90 "a"
              Comment@90..91
                WHITESPACE@90..91 " "
          WHERE@91..96 "where"
          Comment@96..99
            WHITESPACE@96..99 "\n  "
          ClassStatements@99..116
            LAYOUT_START@99..99 ""
            ClassMemberStatement@99..116
              LOWER@99..104 "proof"
              Comment@104..105
                WHITESPACE@104..105 " "
              DOUBLE_COLON@105..107 "::"
              Comment@107..108
                WHITESPACE@107..108 " "
              TypeArrow@108..116
                TypeVariable@108..110
                  LOWER@108..109 "a"
                  Comment@109..110
                    WHITESPACE@109..110 " "
                RIGHT_ARROW@110..112 "->"
                Comment@112..113
                  WHITESPACE@112..113 " "
                TypeVariable@113..116
                  LOWER@113..114 "b"
                  Comment@114..116
                    WHITESPACE@114..116 "\n\n"
            LAYOUT_END@116..116 ""
        LAYOUT_SEPARATOR@116..116 ""
        ClassDeclaration@116..208
          CLASS@116..121 "class"
          Comment@121..122
            WHITESPACE@121..122 " "
          ClassConstraints@122..158
            LEFT_PARENTHESIS@122..123 "("
            TypeApplicationChain@123..137
              TypeConstructor@123..134
                QualifiedName@123..134
                  UPPER@123..133 "TypeEquals"
                  Comment@133..134
                    WHITESPACE@133..134 " "
              TypeVariable@134..136
                LOWER@134..135 "a"
                Comment@135..136
                  WHITESPACE@135..136 " "
              TypeVariable@136..137
                LOWER@136..137 "b"
            COMMA@137..138 ","
            Comment@138..139
              WHITESPACE@138..139 " "
            TypeApplicationChain@139..153
              TypeConstructor@139..150
                QualifiedName@139..150
                  UPPER@139..149 "TypeEquals"
                  Comment@149..150
                    WHITESPACE@149..150 " "
              TypeVariable@150..152
                LOWER@150..151 "b"
                Comment@151..152
                  WHITESPACE@151..152 " "
              TypeVariable@152..153
                LOWER@152..153 "c"
            RIGHT_PARENTHESIS@153..154 ")"
            Comment@154..155
              WHITESPACE@154..155 " "
            LEFT_THICK_ARROW@155..157 "<="
            Comment@157..158
              WHITESPACE@157..158 " "
          ClassHead@158..173
            UPPER@158..168 "TypeEquals"
            Comment@168..169
              WHITESPACE@168..169 " "
            TypeVariable@169..171
              LOWER@169..170 "a"
              Comment@170..171
                WHITESPACE@170..171 " "
            TypeVariable@171..173
              LOWER@171..172 "c"
              Comment@172..173
                WHITESPACE@172..173 " "
          ClassFunctionalDependencies@173..184
            PIPE@173..174 "|"
            Comment@174..175
              WHITESPACE@174..175 " "
            FunctionalDependencyDetermines@175..184
              LOWER@175..176 "a"
              Comment@176..177
                WHITESPACE@176..177 " "
              LOWER@177..178 "b"
              Comment@178..179
                WHITESPACE@178..179 " "
              RIGHT_ARROW@179..181 "->"
              Comment@181..182
                WHITESPACE@181..182 " "
              LOWER@182..183 "c"
              Comment@183..184
                WHITESPACE@183..184 " "
          WHERE@184..189 "where"
          Comment@189..192
            WHITESPACE@189..192 "\n  "
          ClassStatements@192..208
            LAYOUT_START@192..192 ""
            ClassMemberStatement@192..208
              LOWER@192..197 "proof"
              Comment@197..198
                WHITESPACE@197..198 " "
              DOUBLE_COLON@198..200 "::"
              Comment@200..201
                WHITESPACE@200..201 " "
              TypeArrow@201..208
                TypeVariable@201..203
                  LOWER@201..202 "a"
                  Comment@202..203
                    WHITESPACE@202..203 " "
                RIGHT_ARROW@203..205 "->"
                Comment@205..206
                  WHITESPACE@205..206 " "
                TypeVariable@206..208
                  LOWER@206..207 "c"
                  Comment@207..208
                    WHITESPACE@207..208 "\n"
            LAYOUT_END@208..208 ""
      LAYOUT_END@208..208 ""
      END_OF_FILE@208..208 ""
    ,
    [],
)
